pkg/test/framework/components/echo/kube/workload_manager.go

```diff
@@ -0,0 +1,299 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kube
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/go-multierror"
+	kubeCore "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/pkg/config/protocol"
+	echoCommon "istio.io/istio/pkg/test/echo/common"
+	"istio.io/istio/pkg/test/framework/components/echo"
+	"istio.io/istio/pkg/test/framework/components/echo/common"
+	"istio.io/istio/pkg/test/framework/resource"
+	kubeTest "istio.io/istio/pkg/test/kube"
+	"istio.io/istio/pkg/test/util/retry"
+	"istio.io/pkg/log"
+)
+
+var (
+	_ echo.Instance = &instance{}
+	_ io.Closer     = &instance{}
+)
+
+type workloadHandler interface {
+	WorkloadReady(w *workload)
+	WorkloadNotReady(w *workload)
+}
+
+type workloadManager struct {
+	workloads     []*workload
+	mutex         sync.Mutex
+	podController cache.Controller
+	cfg           echo.Config
+	ctx           resource.Context
+	grpcPort      uint16
+	tls           *echoCommon.TLSSettings
+	closing       bool
+	stopCh        chan struct{}
+	handler       workloadHandler
+}
+
+func newWorkloadManager(ctx resource.Context, cfg echo.Config, handler workloadHandler) (*workloadManager, error) {
+	// Get the gRPC port and TLS settings.
+	grpcPort := common.GetPortForProtocol(&cfg, protocol.GRPC)
+	if grpcPort == nil {
+		return nil, errors.New("unable fo find GRPC command port")
+	}
+	var tls *echoCommon.TLSSettings
+	if grpcPort.TLS {
+		tls = cfg.TLSSettings
+	}
+
+	m := &workloadManager{
+		cfg:      cfg,
+		ctx:      ctx,
+		handler:  handler,
+		grpcPort: uint16(grpcPort.InstancePort),
+		tls:      tls,
+		stopCh:   make(chan struct{}, 1),
+	}
+	m.podController = newPodController(cfg, podHandlers{
+		added:   m.onPodAddOrUpdate,
+		updated: m.onPodAddOrUpdate,
+		deleted: m.onPodDeleted,
+	})
+
+	return m, nil
+}
+
+// ReadyWorkloads returns all ready workloads in ascending order by pod name.
+func (m *workloadManager) ReadyWorkloads() ([]echo.Workload, error) {
+	workloads := make([]echo.Workload, 0, len(m.workloads))
+
+	m.mutex.Lock()
+	for _, workload := range m.workloads {
+		if workload.IsReady() {
+			workloads = append(workloads, workload)
+		}
+	}
+	var err error
+	if len(workloads) == 0 {
+		err = fmt.Errorf("no workloads ready for echo %s/%s", m.cfg.Namespace.Name(), m.cfg.Service)
+	}
+	m.mutex.Unlock()
+
+	return workloads, err
+}
+
+func (m *workloadManager) Start() error {
+	log.Warnf("NM: starting workloadManager")
+
+	// Run the pod controller.
+	go m.podController.Run(m.stopCh)
+
+	// Wait until all pods are ready.
+	_, err := retry.Do(func() (result interface{}, completed bool, err error) {
+		// Wait until the cache has synced with the server.
+		if !m.podController.HasSynced() {
```

Comment:
Maybe we should use `cache.WaitForCacheSync(m.stopCh, m.podController.HasSynced)`